drivers/usb/device/usb_dc_usbip.c

```diff
@@ -0,0 +1,922 @@
+}
+
+int usb_dc_ep_mps(const uint8_t ep)
+{
+	uint8_t ep_idx = USB_EP_GET_IDX(ep);
+
+	if (!usbip_ctrl.attached || !usbip_ep_is_valid(ep)) {
+		LOG_ERR("Not attached / wrong endpoint: 0x%x", ep);
+		return -EINVAL;
+	}
+
+	if (USB_EP_DIR_IS_OUT(ep)) {
+		return usbip_ctrl.out_ep_ctrl[ep_idx].mps;
+	} else {
+		return usbip_ctrl.in_ep_ctrl[ep_idx].mps;
+	}
+}
+
+static int usbip_handle_control(struct usbip_command *cmd)
+{
+	uint8_t ep_idx = USB_EP_GET_IDX(cmd->hdr.ep);
+	struct usb_ep_ctrl_prv *ep_ctrl;
+
+	ep_ctrl = &usbip_ctrl.out_ep_ctrl[ep_idx];
+	if (ep_ctrl->cb == NULL) {
+		LOG_ERR("Control endpoint callback not set");
+		return -EIO;
+	}
+
+	if ((cmd->hdr.direction == USBIP_DIR_IN) ^
+	    (REQTYPE_GET_DIR(cmd->submit.bmRequestType) ==
+	     REQTYPE_DIR_TO_HOST)) {
+		LOG_ERR("Failed to verify bmRequestType");
+		return -EIO;
+	}
+
+	ep_ctrl->data_len = 8;
+	LOG_DBG("SETUP event ep 0x%02x %u", ep_idx, ep_ctrl->data_len);
+	memcpy(ep_ctrl->buf, &cmd->submit.bmRequestType, ep_ctrl->data_len);
+	ep_ctrl->cb(ep_idx, USB_DC_EP_SETUP);
+
+	if (cmd->hdr.direction == USBIP_DIR_OUT) {
+		/* Data OUT stage availably */
+		ep_ctrl->data_len = cmd->submit.length;
+		usbip_recv(ep_ctrl->buf, ep_ctrl->data_len);
+		LOG_DBG("DATA OUT event ep 0x%02x %u",
+			ep_idx, ep_ctrl->data_len);
+		ep_ctrl->cb(ep_idx, USB_DC_EP_DATA_OUT);
+	}
+
+	return 0;
+}
+
+static int usbip_handle_data(struct usbip_command *cmd)
+{
+	uint8_t ep_idx = cmd->hdr.ep;
+	struct usb_ep_ctrl_prv *ep_ctrl;
+	uint8_t ep;
+
+	if (cmd->hdr.direction == USBIP_DIR_OUT) {
+		if (ep_idx >= USBIP_OUT_EP_NUM) {
+			return -EINVAL;
+		}
+
+		ep_ctrl = &usbip_ctrl.out_ep_ctrl[ep_idx];
+		ep = ep_idx | USB_EP_DIR_OUT;
+		ep_ctrl->data_len = cmd->submit.length;
+		usbip_recv(ep_ctrl->buf, ep_ctrl->data_len);
+		LOG_DBG("DATA OUT event ep 0x%02x %u", ep, ep_ctrl->data_len);
+
+		ep_ctrl->cb(ep, USB_DC_EP_DATA_OUT);
+
+		/* Send ACK reply */
+		if (!usbip_send_common(ep, 0)) {
+			return -EIO;
+		}
+	} else {
+		if (ep_idx >= USBIP_IN_EP_NUM) {
+			return -EINVAL;
+		}
+
+		ep_ctrl = &usbip_ctrl.in_ep_ctrl[ep_idx];
+		ep = ep_idx | USB_EP_DIR_IN;
+		LOG_DBG("DATA IN event ep 0x%02x %u", ep, ep_ctrl->buf_len);
+
+		/* Send queued data */
+		if (!usbip_send_common(ep, ep_ctrl->buf_len)) {
+			return -EIO;
+		}
+
+		if (usbip_send(ep, ep_ctrl->buf, ep_ctrl->buf_len) !=
+		    ep_ctrl->buf_len) {
+			return -EIO;
+		}
+
+		LOG_HEXDUMP_DBG(ep_ctrl->buf, ep_ctrl->buf_len, ">");
+
+		/*
+		 * Call the callback only if data in usb_dc_ep_write()
+		 * is actually written to the intermediate buffer and sent.
+		 */
+		if (ep_ctrl->buf_len != 0) {
+			ep_ctrl->cb(ep, USB_DC_EP_DATA_IN);
+			usbip_ctrl.in_ep_ctrl[ep_idx].buf_len = 0;
+		}
+	}
+
+	return 0;
+}
+
+static int usbip_recv(uint8_t *buf, size_t len)
+{
+	return recv(usbip_ctrl.connfd, buf, len, MSG_WAITALL);
+}
+
+static int usbip_send(uint8_t ep, const uint8_t *data, size_t len)
+{
+	return send(usbip_ctrl.connfd, data, len, 0);
```

Comment:
It would be better to loop over `send()` here if you treat partial sends as an error condition in the other places in the code. Sending less bytes than requested is not an abnormal TCP socket behaviour.